README.md
# Laravel Simple Select

Laravel Simple Select inputs component for Blade and Livewire.

[![Latest Version on Packagist](https://img.shields.io/packagist/v/victorybiz/laravel-simple-select.svg?style=flat-square)](https://packagist.org/packages/victorybiz/laravel-simple-select)
[![Total Downloads](https://img.shields.io/packagist/dt/victorybiz/laravel-simple-select.svg?style=flat-square)](https://packagist.org/packages/victorybiz/laravel-simple-select)
![GitHub Actions](https://github.com/victorybiz/laravel-simple-select/actions/workflows/main.yml/badge.svg)

### DEMO PREVIEW

![preview](https://github.com/victorybiz/laravel-simple-select/raw/main/demo.gif) 



## Table of Contents
- [Laravel Simple Select](#laravel-simple-select)
    - [DEMO PREVIEW](#demo-preview)
  - [Table of Contents](#table-of-contents)
  - [Installation](#installation)
  - [Requirements](#requirements)
    - [JavaScript Dependencies](#javascript-dependencies)
  - [Usage](#usage)
      - [Simple Select](#simple-select)
      - [Custom Option Slot](#custom-option-slot)
      - [Custom Selected Slot](#custom-selected-slot)
      - [Custom Icon Slots](#custom-icon-slots)
    - [Dependent Selects](#dependent-selects)
    - [Event Listener](#event-listener)
  - [Positioning](#positioning)
  - [Props / Attributes](#props--attributes)
  - [Slots / Custom Display](#slots--custom-display)
  - [Events](#events)
  - [Testing](#testing)
  - [Changelog](#changelog)
  - [Contributing](#contributing)
    - [Security](#security)
  - [Credits](#credits)
  - [License](#license)
  - [Laravel Package Boilerplate](#laravel-package-boilerplate)


<br>

<a name="installation"></a>

## Installation

You can install the package via composer:

```bash
composer require victorybiz/laravel-simple-select
```

**OPTIONAL**: To customize the component, you should publish the configuration file using the vendor:publish Artisan command. The configuration file will be placed in your application's config directory and view file in views directory respectively:

```bash
# Publish the config file
php artisan vendor:publish --tag=simple-select:config
```
```bash
# Publish the view file
php artisan vendor:publish --tag=simple-select:views
```

<a name="requirements"></a>

## Requirements
This package use the following packages.
* Laravel Livewire (https://laravel-livewire.com/) is required when using Livewire `wire:model`
* TailwindCSS (https://tailwindcss.com/) 
* Heroicon (https://heroicons.com/) 
* Alpine.js v3 (https://alpinejs.dev/) 
* Popper.js (https://popper.js.org/)

Please make sure you include these dependencies before using this component. 

<a name="javaScript-dependencies"></a>

### JavaScript Dependencies
For any external JavaScript dependency, we recommend you install them through npm or yarn, and then require them in your project's JavaScript. To install each of the dependencies this package makes use of, run this command in the terminal:
```bash
npm install -D alpinejs @popperjs/core
```
```javascript
import Alpine from 'alpinejs'
import { createPopper } from "@popperjs/core";

window.Alpine = Alpine;
Alpine.start()

window.createPopper = createPopper;
```
If you’re using the compiled JavaScript, don’t forget to include CDN versions of the JavaScript Dependencies before it.

<a name="usage"></a>

## Usage

#### Simple Select
```php
@php
// Basic Arrays
$options = ['Nigeria', 'United Kingdom', 'United States'];
// Above will output Option Value e.g Nigeria 
// Above will output Option Text e.g Nigeria

// OR

// Associative Arrays
$options = [
  ['value' => 'NG', 'text' => 'Nigeria'],
  ['value' => 'GB', 'text' => 'United Kingdom'],
  ['value' => 'US', 'text' => 'United States']
];
// Above will output Option Value e.g NG 
// Above will output Option Text e.g Nigeria

// OR

// Using Associative Arrays data from a Model/Database,
// ensure to customize the field names with value-field="code" and text-field="name" properties of the component.
$options = [
  ['code' => 'NG', 'name' => 'Nigeria'],
  ['code' => 'GB', 'name' => 'United Kingdom'],
  ['code' => 'US', 'name' => 'United States']
];
// OR
$options = [
  ['code' => 'NG', 'name' => 'Nigeria', 'flag' => 'https://www.countryflags.io/ng/shiny/32.png'],
  ['code' => 'GB', 'name' => 'United Kingdom', 'flag' => 'https://www.countryflags.io/gb/shiny/32.png'],
  ['code' => 'US', 'name' => 'United States', 'flag' => 'https://www.countryflags.io/us/shiny/32.png']
];
// Above will output Option Value e.g NG 
// Above will output Option Text e.g Nigeria

@endphp
```
```html
<x-simple-select       
    name="country"
    id="country"
    :options="$options"
    value-field='code'
    text-field='name'
    placeholder="Select Country"
    search-input-placeholder="Search Country"
    :searchable="true"                                               
    class="form-select"     
/>
```

<a name="custom-option-slot"></a>

#### Custom Option Slot
```html
<x-simple-select       
    name="country"
    id="country"
    :options="$options"
    value-field='code'
    text-field='name'
    placeholder="Select Country"
    search-input-placeholder="Search Country"
    :searchable="true"                                               
    class="form-select"     
>
  <x-slot name="customOption">
    <img class="float-left mr-2 -mt-1" :src="option.flag">
    <span x-text="option.name"></span>
  </x-slot>
</x-simple-select>

<x-simple-select       
    name="country"
    id="country"
    :options="$options"
    value-field='code'
    text-field='name'
    placeholder="Select Country"
    search-input-placeholder="Search Country"
    :searchable="true"                                               
    class="form-select"     
>
  <x-slot name="customOption">
    <img class="float-left mr-2 -mt-1" :src="`https://www.countryflags.io/${option.code?.toLowerCase()}/shiny/32.png`">
    <span x-text="option.name"></span>
  </x-slot>
</x-simple-select>
```

<a name="custom-selected-slot"></a>

#### Custom Selected Slot
```html
<x-simple-select       
    name="country"
    id="country"
    :options="$options"
    value-field='code'
    text-field='name'
    placeholder="Select Country"
    search-input-placeholder="Search Country"
    :searchable="true"                                               
    class="form-select"     
>
  <x-slot name="customSelected">
    <img class="float-left mr-2" :src="`https://www.countryflags.io/${option.code?.toLowerCase()}/shiny/24.png`">
    <span x-text="option.name"></span>
  </x-slot>
  <x-slot name="customOption">
    <img class="float-left mr-2 -mt-1" :src="`https://www.countryflags.io/${option.code?.toLowerCase()}/shiny/32.png`">
    <span x-text="option.name"></span>
  </x-slot>
</x-simple-select>
```


<a name="custom-icon-slot"></a>

#### Custom Icon Slots
```html
<x-simple-select       
    name="country"
    id="country"
    :options="$options"
    value-field='code'
    text-field='name'
    placeholder="Select Country"
    search-input-placeholder="Search Country"
    :searchable="true"                                               
    class="form-select"     
>
  <x-slot name="customOption">
    <img class="float-left mr-2 -mt-1" :src="option.flag">
    <span x-text="option.name"></span>
  </x-slot>
  <x-slot name="customDeselectOptionIcon">
    <!-- Heroicon solid/x-circle -->
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" class = 'h-4 fill-current'>
      <path d="M12 2c5.514 0 10 4.486 10 10s-4.486 10-10 10-10-4.486-10-10 4.486-10 10-10zm0-2c-6.627 0-12 5.373-12 12s5.373 12 12 12 12-5.373 12-12-5.373-12-12-12zm6 16.538l-4.592-4.548 4.546-4.587-1.416-1.403-4.545 4.589-4.588-4.543-1.405 1.405 4.593 4.552-4.547 4.592 1.405 1.405 4.555-4.596 4.591 4.55 1.403-1.416z"/>
    </svg>
  </x-slot>
</x-simple-select>
```

<a name="dependent-selects"></a>

### Dependent Selects
If you have a custom select whose options depend on the selection of another select, or just some kind of condition to be met, you can listen to the updated event of the livewire model of the main select to update the options in the dependent select.
```php
// Expected data in Database
// Model Country::class 
$countries = [
  ['code' => 'NG', 'name' => 'Nigeria'],
  ['code' => 'GB', 'name' => 'United Kingdom'],
  ['code' => 'US', 'name' => 'United States']
];
// Model State::class
$states = [
  ['id' => 1, 'country_code' => 'NG', 'name' => 'Abuja'],
  ['id' => 2, 'country_code' => 'NG', 'name' => 'Edo'],
  ['id' => 3, 'country_code' => 'NG', 'name' => 'Lagos'],
  ['id' => 4, 'country_code' => 'US', 'name' => 'Alaska'],
  ['id' => 5, 'country_code' => 'US', 'name' => 'Califonia'],
  ['id' => 6, 'country_code' => 'US', 'name' => 'Florida'],
  ['id' => 7, 'country_code' => 'GB', 'name' => 'Belfast'],
  ['id' => 8, 'country_code' => 'GB', 'name' => 'London'],
  // ...
];
```

Create a livewire component as the form page
```php
<?php

namespace App\Http\Livewire;

use Livewire\Component;

class CreateUser extends Component
{
    public $countries = [];
    public $states = [];

    public $name;
    public $country;
    public $state;

    protected function rules()
    {
        // 
    }

    public function updated($propertyName)
    {
        $this->validateOnly($propertyName);
    }

    public function store()
    {
        $this->validate();
        // Store the data
    }

    public function mount()
    {
        $this->countries = \App\Models\Country::orderBy('name')->get()->toArray();             
    }

    public function updatedCountry($countryCode)
    {   
        if ($countryCode) {
            $this->states = \App\Models\State::where('country_code', $countryCode)->orderBy('name')->get()->toArray();  
        } else {
            $this->states = [];            
        }   
        $this->state = null;
    }

    public function render()
    {
        return view('livewire.create-user');
    }
}

```
In your component view
```html
 <form wire:submit.prevent="store">

  <label for="name">Name</label>
  <div class="mt-1">
    <input       
        wire:model="name"
        name="name"
        id="name"
        placeholder="Enter name"                                            
        class="form-input"     
    />
  </div>

  <label for="country">Country</label>
  <div class="mt-1">
    <x-simple-select       
        wire:model="country"
        name="country"
        id="country"
        :options="$options"
        value-field='code'
        text-field='name'
        placeholder="Select Country"
        search-input-placeholder="Search Country"
        :searchable="true"                                               
        class="form-select"     
    />
  </div>

  <label for="state">State</label>
  <div class="mt-1">
    <x-simple-select       
        wire:model="state"
        name="state"
        id="state"
        :options="$states"
        value-field='id'
        text-field='name'
        placeholder="Select State"
        search-input-placeholder="Search State"
        :searchable="true"                       
        class="form-select"
    />
  </div>
</form>
```
<a name="event-listener"></a>

### Event Listener
```javascript
window.addEventListener('select', function(option) {
    console.log(option.detail.value); // Select option value(s)
    console.log(option.detail.name); // The select element name
    console.log(option.detail.id); // The select element ID
});
```

<a name="positioning"></a>

## Positioning

The simple-select component makes use of `Popper.js` for positioning the select menu. This should remove the need for fixed positioning the select menu now. In addition to positioning the menu when opened, Popper.js will also re-position the menu as needed when the page is scrolled.

<a name="props"></a>

## Props / Attributes

| Name | Type | Default | Required | Description |
| --- | --- | --- | --- | --- |
| **id** | `Integer\|\|String` | | Yes | Used to identify the component in events. |
| **name** | `Integer\|\|String` | | Yes | Specifies a name for component. |
| **options** | `Array` | | Yes | Array of available options: Objects, Strings or Integers. If array of objects, visible text/label will default to `option.text` and value default to `option.value`. |
| **value-field** | `String` | `'value'` | No | Array key for option value if `options` is an associative array. |
| **text-field** | `String` | `'text'` | No | Array key for option text if `options` is an associative array. |
| **value** | `Array\|\|String\|\|Integer` | `null` | No | Presets the selected options. |
| **placeholder** | `String` | `'Select Option'` | No | Equivalent to the `placeholder` attribute on a `<select>` input. | 
| **searchable** | `Boolean` | `true` | No | Show / hide options search input. | 
| **search-input-placeholder** | `String` | `'Search...'` | No | Equivalent to the `placeholder` attribute on a `<input>`. | 
| **clearable** | `Boolean` | `false` | No | Enable support for clearable selection. Use only for a non `multiple` select. | 
| **class** | `String` |  | No | Equivalent to the `class` attribute on a `<select>` input. | 
| **multiple** | `Boolean` | `false` | No | Equivalent to the `multiple` attribute on a `<select>` input. This also enable multiple options tagging if set. | 
| **max-selection** | `Integer` | | No | Limit number of allowed selected options. | 
| **required** | `Boolean` | `false` | No | Equivalent to the `required` attribute on a `<select>` input. | 
| **disabled** | `Boolean` | `false` | No | Equivalent to the `disabled` attribute on a `<select>` input. | 
| **no-options** | `String` | `'No option data.'` | No | Message to show when options list is empty.|
| **no-result** | `String` | `'No results match your search.'` | No | Message to show when no option.|
| **on-select** | `String` | `'select'` | No | Customize event name of an event emitted after selecting an option. |

<a name="slots"></a>

## Slots / Custom Display

| Name | Description |
| --- | --- |
| **customOption** | Slot for custom option text template. See [example](#custom-option-slot) above. |
| **customSelected** | Slot for custom selected template. See [example](#custom-selected-slot) above. |
| **customDeselectOptionIcon**| Slot for custom deselect option icon markup. See [example](#custom-icon-slot) above. |
| **customCaretDownIcon**| Slot for custom caret down icon markup. See [example](#custom-icon-slot) above. |
| **customCaretUpIcon**| Slot for custom caret up icon markup. See [example](#custom-icon-slot) above. |

<a name="events"></a>

## Events

| Name | Listen to |  Description |
| --- | --- |  --- |
| **Select** | `select` | Emitted after selecting an option. See [example](#event-listener) above. |


## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security


CHANGELOG.md
# Changelog

All notable changes to `laravel-simple-select` will be documented in this file

## v1.5.0 - 2023-02-14

- Added support for Laravel 10 by @laravel-shift in #14
- 
## v1.4.0 - 2022-02-09

- Added support for Laravel 9 by @dcblogdev in #7
  
## v1.3.0 - 2022-01-29

- Added TailwindCSS Dark Mode support.
  
## v1.2.5 - 2021-11-21

- Fixed bugs in usage of integer as field value.
  
## v1.2.4 - 2021-11-12

- Added default value for `id` prop.
  
## v1.2.3 - 2021-08-11

- `customSelected` Slot Bug fix.
  
## v1.2.2 - 2021-08-04

- Popper positioning bug fix
  
## v1.2.1 - 2021-07-26

- wire:model and popper bug fix.
  
## v1.2.0 - 2021-06-26

- Added `customSelected` Slot.
  
## v1.1.2 - 2021-06-26

- Fix bugs in `:clearable` props.
  
## v1.1.1 - 2021-06-26

- Added `:clearable` prop.

## v1.0.6 - 2021-06-26

- Updated `README.md`

## v1.0.5 - 2021-06-25

- Fix bugs in :value props.
  
## v1.0.4 - 2021-06-25

- Added demo.gif image preview to  `README.md`

## v1.0.3 - 2021-06-25

- Bug fix.

## v1.0.2 - 2021-06-25

- Bug fix.

## v1.0.1 - 2021-06-25

- Bug fix.

## v1.0.0 - 2021-06-25

- Major release

## v0.1.1 - 2021-06-24

- initial release bug fix.
  
## v0.1.0 - 2021-06-24

- initial release


resources/views/components/select.blade.php
<div
    class="relative mt-1"
    x-data="Select({
        dataSource: {{ json_encode($options) }},
        selected: @entangle($attributes->wire('model')),
        placeholder: '{{ $placeholder }}',
        valueField: '{{ $valueField }}',
        textField: '{{ $textField }}',
        name: '{{ $name }}',
        id: '{{ $id }}',
        searchPlaceholder: '{{ $searchPlaceholder }}',
        emptyOptionsMessage: '{{ $emptyOptionsMessage }}',
        emptyOptionsMessageAfterSearch: '{{ $emptyOptionsMessageAfterSearch }}',
        multiple: {{ isset($attributes['multiple']) ? 'true' : 'false' }},
        maxSelection: '{{ $maxSelection }}',
        disabled: {{ isset($attributes['disabled']) ? 'true' : 'false' }},
        searchable: {{ $searchable ? 'true' : 'false' }},
    })"
    x-init="init();"
    x-on:click.outside="closeSelect()"
    x-on:keydown.escape="closeSelect()"
    :wire:key="name ?? generateID()"
>
    <div
        x-ref="simpleLivewireSelectButton"
        x-on:click="toggleSelect()"
        x-on:keyup.enter="toggleSelect()"
        tabindex="0"
        x-bind:class="{
            'rounded-md': !open,
            'rounded-t-md': open,
            'bg-gray-200 cursor-default': disabled
        }"
        {{ $attributes->class('block w-full border border-gray-300 rounded-md shadow-sm focus:ring-0 focus:ring-gray-400 focus:border-gray-400 sm:text-sm sm:leading-5')->only('class'); }}
        
    > 
        <div x-cloak x-show="!selected || selected.length === 0" class="flex flex-wrap">
            <div class="text-gray-800 rounded-sm w-full truncate px-2 py-0.5 my-0.5 flex flex-row items-center">
                <div class="w-full px-2 truncate" x-text="placeholder">&nbsp;</div>
                <div x-show="!disabled" x-bind:class="{ 'cursor-pointer': !disabled }" class="h-4" x-on:click.prevent.stop="toggleSelect()">
                    <span x-show="!open">
                        @isset($customToggleDownIcon)
                            {{ $customToggleDownIcon }}
                        @else
                            {{-- Heroicon: outline/chevron-down --}}
                            <svg xmlns="http://www.w3.org/2000/svg" class="w-6 h-6 text-gray-400" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                                <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M19 9l-7 7-7-7" />
                            </svg>
                        @endisset 
                    </span>
                    <span x-show="open">
                        @isset($customToggleUpIcon)
                            {{ $customToggleUpIcon }}
                        @else
                            {{-- Heroicon: outline/chevron-up --}}
                            <svg xmlns="http://www.w3.org/2000/svg" class="w-6 h-6 text-gray-400" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                                <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M5 15l7-7 7 7" />
                            </svg>
                        @endisset 
                    </span>
                </div>
            </div>
        </div>
        @isset($attributes['multiple'])
            <div x-cloak x-show="selected != null && typeof selected === 'object' && selected.length > 0" class="flex flex-wrap space-x-1">
                <template x-for="(value, index) in selected" :key="index">
                    <div class="text-gray-800 rounded-full truncate bg-gray-300 px-2 py-0.5 my-0.5 flex flex-row items-center">
                        <div class="px-2 truncate" x-text="getTextFromSelectedValue(value)"></div>
                        <div
                            x-show="!disabled"
                            x-bind:class="{ 'cursor-pointer': !disabled }"
                            x-on:click.prevent.stop="deselectOption(index)"
                            x-on:keyup.enter="deselectOption(index)"
                            class="w-4"
                            tabindex="0"
                        >
                            @isset($customDeselectOptionIcon)
                                {{ $customDeselectOptionIcon }}
                            @else
                                {{-- Heroicon solid/x-circle --}}
                                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" class = 'h-4 fill-current'>
                                    <path d="M12 2c5.514 0 10 4.486 10 10s-4.486 10-10 10-10-4.486-10-10 4.486-10 10-10zm0-2c-6.627 0-12 5.373-12 12s5.373 12 12 12 12-5.373 12-12-5.373-12-12-12zm6 16.538l-4.592-4.548 4.546-4.587-1.416-1.403-4.545 4.589-4.588-4.543-1.405 1.405 4.593 4.552-4.547 4.592 1.405 1.405 4.555-4.596 4.591 4.55 1.403-1.416z"/>
                                </svg>
                            @endisset 
                        </div>
                    </div>
                </template>
            </div>
        @else            
            <div x-cloak x-show="selected" class="flex flex-wrap"> 
                <div class="text-gray-800 rounded-sm w-full truncate px-2 py-0.5 my-0.5 flex flex-row items-center">
                    <div class="w-full px-2 truncate" x-text="getTextFromSelectedValue(selected)"></div>
                    <div
                        x-show="!disabled"
                        x-bind:class="{ 'cursor-pointer': !disabled }"
                        x-on:click.prevent.stop="deselectOption()"
                        x-on:keyup.enter="deselectOption()"
                        class="h-4"
                        tabindex="0"
                    >
                        @isset($customDeselectOptionIcon)
                            {{ $customDeselectOptionIcon }}
                        @else
                            {{-- Heroicon solid/x-circle --}}
                            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" class = 'h-4 fill-current'>
                                <path d="M12 2c5.514 0 10 4.486 10 10s-4.486 10-10 10-10-4.486-10-10 4.486-10 10-10zm0-2c-6.627 0-12 5.373-12 12s5.373 12 12 12 12-5.373 12-12-5.373-12-12-12zm6 16.538l-4.592-4.548 4.546-4.587-1.416-1.403-4.545 4.589-4.588-4.543-1.405 1.405 4.593 4.552-4.547 4.592 1.405 1.405 4.555-4.596 4.591 4.55 1.403-1.416z"/>
                            </svg>
                        @endisset                        
                    </div>
                </div>
            </div>
        @endisset
    </div>
    {{-- style="height: 17.2rem;" --}}
    <div x-ref="simpleLivewireSelectOptionsContainer" x-bind:style="open ? 'height: ' + popperHeight : ''" class="absolute w-full">
        <div x-show="open">
            <input
                type="search"
                x-show="searchable"
                x-ref="simpleLivewireSelectOptionsSearch"
                x-model="search"
                x-on:click.prevent.stop="open=true"
                :placeholder="searchPlaceholder"
                class="block w-full p-2 bg-gray-100 border border-gray-300 shadow-md focus:border-gray-200 focus:ring-0 sm:text-sm sm:leading-5"
            />
            <ul                
                x-ref="simpleLivewireSelectOptionsList"
                class="absolute z-10 w-full py-1 overflow-auto text-base bg-white shadow-lg rounded-b-md max-h-60 ring-1 ring-black ring-opacity-5 focus:outline-none sm:text-sm"
                tabindex="-1"
                role="listbox"
            >
                <div x-cloak x-show="Object.values(options).length == 0 && search.toString().trim() == ''" x-text="emptyOptionsMessage" class="px-2 py-2">&nbsp;</div>
                <div x-cloak x-show="Object.values(options).length == 0 && search.toString().trim() != ''" x-text="emptyOptionsMessageAfterSearch" class="px-2 py-2">&nbsp;</div>
                <template x-for="(option, index) in Object.values(options)" :key="index">
                    <li               
                        :tabindex="index"             
                        class="relative py-2 pl-3 select-none pr-9 "
                        @isset($attributes['multiple'])
                            x-bind:class="{
                                'bg-gray-300 text-black hover:none': selected && selected.includes(getOptionValue(option, index)),
                                'text-gray-900 cursor-defaul hover:bg-gray-200 hover:cursor-pointer focus:bg-gray-200': !(selected && selected.includes(getOptionValue(option, index))),
                            }"
                        @else
                            x-bind:class="{
                                'bg-gray-300 text-black hover:none': selected == getOptionValue(option, index),
                                'text-gray-900 cursor-defaul hover:bg-gray-200 hover:cursor-pointer focus:bg-gray-200': !(selected == getOptionValue(option, index)),
                            }"
                        @endisset
                        x-on:click="selectOption(getOptionValue(option, index))"
                        x-on:keyup.enter="selectOption(getOptionValue(option, index))"
                    >
                        @isset($customOption)
                            {{ $customOption }}
                        @else
                            <span x-text="getOptionText(option, index)"></span>
                        @endisset
                    </li>
                </template>
            </ul>
        </div>
    </div>
</div>

<script>
    function Select(config) {
        return {
            open: false,
            search: '',
            dataSource: config.dataSource,
            options: {},
            valueField: config.valueField,
            textField: config.textField,
            placeholder: config.placeholder,
            selected: config.selected,
            searchable: config.searchable,
            disabled: config.disabled,
            multiple: config.multiple,
            maxSelection: config.maxSelection,
            name: config.name,
            id: config.id,
            searchPlaceholder: config.searchPlaceholder,
            emptyOptionsMessage: config.emptyOptionsMessage,
            emptyOptionsMessageAfterSearch: config.emptyOptionsMessageAfterSearch,
            isLoading: false,
            popperInstance: null,
            popperHeight: '0px', // 17.2rem

            init: function() {
                if(!this.selected || this.selected == null){
                    if (this.multiple) {
                        this.selected = []
                    } else {
                        this.selected = ''
                    }
                }
                this.resetOptions();

                // this.$watch('dataSource', ((dataSource) => {
                //     if(!this.selected || this.selected == null){
                //         if (this.multiple) {
                //             this.selected = []
                //         } else {
                //             this.selected = ''
                //         }
                //     }
                //     setTimeout(() =>  this.resetOptions(dataSource), 1000)
                //     // this.resetOptions(dataSource);
                // }));

                this.$watch('search', ((search) => {
                    this.resetOptions();
                    this.options = Object.values(this.dataSource)
                        .filter((value, index) => this.getOptionText(value, index).toString().toLowerCase().includes(search.toLowerCase().trim()));
                    
                    setTimeout(() => {
                        this.popper();
                        this.scrollToOption();
                    }, 100);
                }));
            },

            generateID: () => {
                return '_' + Math.random().toString(36).substr(2, 9)
            },
            
            resetOptions: function(dataSource = null) {                
                if (!dataSource) {
                    dataSource = this.dataSource;
                }
                this.options = Object.values(dataSource);       
            },

            toggleSelect: function() {
                if (!this.disabled) {
                    if (this.open) {
                        this.closeSelect();
                    } else {
                        this.open = true;
                                                
                        setTimeout(() => {
                            this.popper();
                            this.scrollToOption();      
                            this.checkMaxSelectionReached();  
                        }, 700);          
                    }                      
                }
            },

            closeSelect: function() {
                this.open = false;
                this.search = '';
            },

            selectOption: function(value) {
                if(!this.disabled) {
                    // If multiple push to the array, if not, keep that value and close menu
                    if (this.multiple) {
                        // If it's not already in there
                        if (!this.selected.includes(value)) {
                            if (this.maxSelection == 0 || (this.maxSelection > 0 && this.selected.length < this.maxSelection)) {
                                this.selected.push(value)                                
                            }
                            let reached = this.checkMaxSelectionReached();
                            if (reached) {
                                this.closeSelect();
                            }
                        }
                    } else {
                        this.selected = value;
                        this.closeSelect();
                    }
                }
            },
            
            deselectOption: function(index = null) {
                if (this.multiple) {
                    this.selected.splice(index, 1)
                } else {
                    this.selected = ''
                }
            },

            checkMaxSelectionReached: function() {
                if (this.multiple && this.$refs.simpleLivewireSelectOptionsList) {
                    if (this.maxSelection > 0 && this.selected.length >= this.maxSelection) {
                        this.$refs.simpleLivewireSelectOptionsList.querySelectorAll(':scope > li').forEach((el) => {
                            el.style.pointerEvents = 'none';
                            el.style.opacity = '0.4';
                        });
                        return true;
                    } else {
                        this.$refs.simpleLivewireSelectOptionsList.querySelectorAll(':scope > li').forEach((el) => {
                            el.style.pointerEvents = 'auto';
                            el.style.opacity = '1';
                        });
                    }
                }
                return false;
            },

            getOptionValue: function(option, index) {
                return typeof option === 'object' && this.options[index] ? this.options[index][this.valueField] : option;
            },

            getOptionText: function(option, index) {
                return typeof option === 'object' && this.options[index] ? this.options[index][this.textField] : option;
            },

            getIndexFromSelectedValue: function(value) {    
                let valueField = this.valueField;
                return Object.values(this.dataSource).findIndex(function(x) {
                    if (typeof x === 'object') {
                        return x[valueField] === value;
                    } else {
                        return x === value;
                    }
                });       
            },

            getTextFromSelectedValue: function(value) {
                let index = this.getIndexFromSelectedValue(value);
                let valueField = this.valueField;
                let foundValue = Object.values(this.dataSource).find(function(x) {
                    if (typeof x === 'object') {
                        return x[valueField] === value;
                    } else {
                        return x === value;
                    }
                });
                return typeof foundValue === 'object' && this.dataSource[index] ? this.dataSource[index][this.textField] : foundValue;
            },

            popper: function() {
                // update popper position
                if (this.$refs.simpleLivewireSelectOptionsList && this.$refs.simpleLivewireSelectOptionsList.offsetHeight) {
                    this.popperHeight = (this.$refs.simpleLivewireSelectOptionsList.offsetHeight + 20) + 'px';
                }                
                            
                if (window.createPopper && this.$refs.simpleLivewireSelectButton && this.$refs.simpleLivewireSelectOptionsContainer) {
                    this.popperInstance = createPopper(this.$refs.simpleLivewireSelectButton, this.$refs.simpleLivewireSelectOptionsContainer, {
                        placement: "bottom-start",
                        modifiers: [
                            // {
                            //     name: 'offset',
                            //     options: {
                            //         offset: [0, 8],
                            //     },
                            // },
                            {
                                name: "preventOverflow",
                                options: { 
                                    boundary: "clippingParents" 
                                },
                            },
                            { name: "flip", options: { padding: 50 } },
                        ]
                    });
                }
            },

            scrollToOption: function () {                
                try {
                    if (this.selected && this.$refs.simpleLivewireSelectOptionsList) {
                        let focusIndex = 0;
                        if (this.multiple) {
                            let lastSelected = this.selected.length > 0 ? this.selected[this.selected.length - 1] : '';
                            focusIndex = lastSelected ? this.getIndexFromSelectedValue(lastSelected) : 0;
                        } else {
                            focusIndex = this.getIndexFromSelectedValue(this.selected);
                        }                        
                        // let nonListItem = 3;
                        // let totalListItem = this.$refs.simpleLivewireSelectOptionsList.children.length > nonListItem ? this.$refs.simpleLivewireSelectOptionsList.children.length - nonListItem : 0;
                        let optionsList = this.$refs.simpleLivewireSelectOptionsList.querySelectorAll(':scope > li');
                        let totalOptionsList = optionsList.length;
                        if (totalOptionsList > 0) {
                            let offsetTop = optionsList[focusIndex].offsetTop;
                            this.$refs.simpleLivewireSelectOptionsList.scrollTop = offsetTop || 0;
                            // optionsList[focusIndex].focus();                            
                        } else {
                            this.$refs.simpleLivewireSelectOptionsList.scrollTop = 0;                            
                        }                        
                    } else {
                        this.$refs.simpleLivewireSelectOptionsList.scrollTop = 0;
                    }
                    this.$refs.simpleLivewireSelectOptionsSearch.focus();
                } catch (e) {}                    
                
            }
        }
    }
    window.Select = Select;
</script>